##### ARTI 560 - Computer Vision

## Image Classification with Vision Transformer (ViT) - Exercise

### Objective

In this exercise, you will test the pretrained Vision Transformer (ViT) model on 5 real-world images that you find online.

You will:

1. Download 5 images for different classes in [ImageNet](https://github.com/Waikato/wekaDeeplearning4j/blob/master/docs/user-guide/class-maps/IMAGENET.md).

2. Load the ImageNet class names from a [text file](https://raw.githubusercontent.com/pytorch/hub/master/imagenet_classes.txt).

3. Use ViT to predict the class for each image.

4. Record whether the prediction was correct.

#### Important Note

For this exercise, you MUST use the following KerasHub components:

- [keras_hub.models.ViTImageClassifier](https://keras.io/keras_hub/api/models/vit/vit_image_classifier/)

- [keras_hub.models.ViTImageClassifierPreprocessor](https://keras.io/keras_hub/api/models/vit/vit_image_classifier_preprocessor/)

This ensures your input preprocessing (resizing + normalization) matches what the pretrained ViT model expects.

Do not replace the preprocessor with manual normalization (such as dividing by 255), because it may produce incorrect predictions.

In [20]:
# Import Libraries
import requests
import numpy as np
import keras
import keras_hub

# Load ImageNet class names
classes_url = "https://raw.githubusercontent.com/pytorch/hub/master/imagenet_classes.txt"
imagenet_classes = requests.get(classes_url).text.splitlines()

# Load ViTImageClassifierPreprocessor (vit_base_patch16_224_imagenet preset)
preprocessor = keras_hub.models.ViTImageClassifierPreprocessor.from_preset(
    "vit_base_patch16_224_imagenet"
)

# Load ViTImageClassifier (vit_base_patch16_224_imagenet preset)
classifier = keras_hub.models.ViTImageClassifier.from_preset(
    "vit_base_patch16_224_imagenet",
    preprocessor=preprocessor,
    activation="softmax"
)

print("--- ViT Predictions ---\n")

# Load the images
image_paths = {
    "Snail": "/content/snail_.png",
    "Jellyfish": "/content/jelly_fish.png",
    "Goldfish": "/content/gold_fish.png",
    "Tree Frog": "/content/tree_frog.png",
    "Water Snake": "/content/water_snake.png"
}

for i, (name, path) in enumerate(image_paths.items()):
    img = keras.utils.load_img(path)
    img_array = keras.utils.img_to_array(img)
    img_batch = np.expand_dims(img_array, axis=0)

    # Predict classes
    pred = classifier.predict(img_batch, verbose=0)

    best_index = np.argmax(pred[0])
    class_name = imagenet_classes[best_index]
    confidence = pred[0][best_index] * 100

    print("Image",i+1,": ", name)
    print("Prediction: ",class_name)
    print(f"Confidence: {confidence:.2f}%")
    print("\n")

--- ViT Predictions ---

Image 1 :  Snail
Prediction:  snail
Confidence: 99.52%


Image 2 :  Jellyfish
Prediction:  jellyfish
Confidence: 99.95%


Image 3 :  Goldfish
Prediction:  goldfish
Confidence: 99.97%


Image 4 :  Tree Frog
Prediction:  tree frog
Confidence: 91.01%


Image 5 :  Water Snake
Prediction:  water snake
Confidence: 77.67%




### Record Your Results

Fill the table below based on your results:

| Image File   | Predicted Label | True Label (What you searched) | Correct? (Yes/No) |
| ------------ | --------------- | ------------------------------ | ----------------- |
|  snail_.png            | snail (Confidence: 99.52%)          |     Snail                           | YES            |
|   jelly_fish.png         | jellyfish (Confidence: 99.95%)          |         Jellyfish                       | YES            |
|     gold_fish.png         | goldfish (Confidence: 99.97%)          |          Goldfish                      | YES            |
|       tree_frog.png       | tree frog (Confidence: 91.01%)          |         Tree Frog                       | YES            |
|         water_snake.png     | water snake (Confidence: 77.67%)          |        Water Snake                        | YES            |
